File: typescript/src/Base58Codec.ts
const BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
const BASE58_INDEXES = createBase58Indexes();

function createBase58Indexes(): Int16Array {
    const indexes = new Int16Array(128);
    indexes.fill(-1);
    for (let index = 0, length = BASE58_ALPHABET.length; index < length; index++) {
        indexes[BASE58_ALPHABET.charCodeAt(index)] = index;
    }
    return indexes;
}

export function encodeBase58(bytes: Uint8Array): string {
    if (bytes.length === 0) {
        return "";
    }

    let leadingZeroCount = 0;
    while (leadingZeroCount < bytes.length && bytes[leadingZeroCount] === 0) {
        leadingZeroCount++;
    }

    if (leadingZeroCount === bytes.length) {
        return "1".repeat(leadingZeroCount);
    }

    const digits: number[] = [0];
    for (let byteIndex = leadingZeroCount, byteLength = bytes.length; byteIndex < byteLength; byteIndex++) {
        let carry = bytes[byteIndex] ?? 0;
        for (let digitIndex = 0, digitLength = digits.length; digitIndex < digitLength; digitIndex++) {
            carry += (digits[digitIndex] ?? 0) * 256;
            digits[digitIndex] = carry % 58;
            carry = Math.floor(carry / 58);
        }
        while (carry > 0) {
            digits.push(carry % 58);
            carry = Math.floor(carry / 58);
        }
    }

    let encoded = "1".repeat(leadingZeroCount);
    for (let digitIndex = digits.length - 1; digitIndex >= 0; digitIndex--) {
        encoded += BASE58_ALPHABET[digits[digitIndex] ?? 0] ?? "";
    }
    return encoded;
}

export function decodeBase58(encoded: string): Uint8Array {
    if (encoded.length === 0) {
        return new Uint8Array(0);
    }

    let leadingOneCount = 0;
    while (leadingOneCount < encoded.length && encoded.charCodeAt(leadingOneCount) === 49) {
        leadingOneCount++;
    }

    if (leadingOneCount === encoded.length) {
        return new Uint8Array(leadingOneCount);
    }

    const bytes: number[] = [0];
    for (
        let characterIndex = leadingOneCount, characterLength = encoded.length;
        characterIndex < characterLength;
        characterIndex++
    ) {
        const characterCode = encoded.charCodeAt(characterIndex);
        const alphabetIndex = characterCode < BASE58_INDEXES.length ? BASE58_INDEXES[characterCode] : -1;
        if (alphabetIndex === undefined || alphabetIndex < 0) {
            throw new Error(`Invalid base58 character at index ${characterIndex}`);
        }

        let carry = alphabetIndex;
        for (let byteIndex = 0, byteLength = bytes.length; byteIndex < byteLength; byteIndex++) {
            carry += (bytes[byteIndex] ?? 0) * 58;
            bytes[byteIndex] = carry & 0xff;
            carry >>= 8;
        }
        while (carry > 0) {
            bytes.push(carry & 0xff);
            carry >>= 8;
        }
    }

    const decoded = new Uint8Array(leadingOneCount + bytes.length);
    for (let byteIndex = 0, byteLength = bytes.length; byteIndex < byteLength; byteIndex++) {
        decoded[decoded.length - 1 - byteIndex] = bytes[byteIndex] ?? 0;
    }
    return decoded;
}

export function decodePublicKey(publicKey: string | Uint8Array): Uint8Array {
    const bytes = typeof publicKey === "string" ? decodeBase58(publicKey) : new Uint8Array(publicKey);
    if (bytes.length !== 32) {
        throw new Error(`Public key must contain 32 bytes; received ${bytes.length}`);
    }
    return bytes;
}

export function normalizePublicKey(publicKey: string | Uint8Array): string {
    return encodeBase58(decodePublicKey(publicKey));
}

File: python/chancery_reference/base58_codec.py
from __future__ import annotations

BASE58_ALPHABET = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"
BASE58_INDEXES = {character: index for index, character in enumerate(BASE58_ALPHABET)}


def encode_base58(data: bytes | bytearray | memoryview) -> str:
    raw = bytes(data)
    if not raw:
        return ""

    leading_zero_count = 0
    while leading_zero_count < len(raw) and raw[leading_zero_count] == 0:
        leading_zero_count += 1

    if leading_zero_count == len(raw):
        return "1" * leading_zero_count

    digits = [0]
    for byte_value in raw[leading_zero_count:]:
        carry = byte_value
        for digit_index in range(len(digits)):
            carry += digits[digit_index] * 256
            digits[digit_index] = carry % 58
            carry //= 58
        while carry > 0:
            digits.append(carry % 58)
            carry //= 58

    encoded = "1" * leading_zero_count
    encoded += "".join(BASE58_ALPHABET[digit] for digit in reversed(digits))
    return encoded


def decode_base58(encoded: str) -> bytes:
    if not encoded:
        return b""

    leading_one_count = 0
    while leading_one_count < len(encoded) and encoded[leading_one_count] == "1":
        leading_one_count += 1

    if leading_one_count == len(encoded):
        return b"\x00" * leading_one_count

    decoded = [0]
    for character_index, character in enumerate(encoded[leading_one_count:], start=leading_one_count):
        alphabet_index = BASE58_INDEXES.get(character)
        if alphabet_index is None:
            raise ValueError(f"Invalid base58 character at index {character_index}")
        carry = alphabet_index
        for byte_index in range(len(decoded)):
            carry += decoded[byte_index] * 58
            decoded[byte_index] = carry & 0xFF
            carry >>= 8
        while carry > 0:
            decoded.append(carry & 0xFF)
            carry >>= 8

    return b"\x00" * leading_one_count + bytes(reversed(decoded))


def decode_public_key(public_key: str | bytes | bytearray | memoryview) -> bytes:
    data = decode_base58(public_key) if isinstance(public_key, str) else bytes(public_key)
    if len(data) != 32:
        raise ValueError(f"Public key must contain 32 bytes; received {len(data)}")
    return data


def normalize_public_key(public_key: str | bytes | bytearray | memoryview) -> str:
    return encode_base58(decode_public_key(public_key))
